Better gcd: Lehmers algorithm

jzmaddock opened the pull request.

NAThompson commented:
@jzmaddock : Commit the benchmarks too!

jzmaddock commented:
> Commit the benchmarks too!

:)  Getting there, done now.

Still a work in progress, but here's the current results for msvc, scores are relative to GMP assigned a score of 1 on each row.


Bits | Previous | PR
-- | -- | --
  |   |  
1024 | 3.9398084234 | 2.4983388815
2048 | 5.161766933 | 2.695700359
4096 | 6.8414634574 | 2.7500000172
8192 | 8.380952381 | 2.9047619048
16384 | 10.5396825397 | 2.9841269841
32678 | 11.5585585586 | 3

NAThompson commented:
@jzmaddock : I changed the units on your benchmarks to make it not print so many insignificant digits. BTW, is there a reason why you are running the `BM_gcd_old` twice?

https://github.com/boostorg/multiprecision/blob/86b00c2f48a17854dd4f8b11e73a46b9789e1e9e/performance/gcd_bench.cpp#L207-L209

jzmaddock commented:
> @NAThompson :BTW, is there a reason why you are running the BM_gcd_old twice?

I was getting screwy/slow results from whatever run went first, so it was just a safeguard against deluding myself ;)

BTW I have a double-limb version working which makes a huge difference on msvc, but is marginally slower for gcc since it involves `__int128` arithmetic.  The next job is start looking at which arithmetic operations can be streamlined.

With regard to the division: In Jebelean's paper on the GMP version, he notes that since the quotient is usually small, stepwise subtraction to find the quotient is often faster, falling back to division only when it's clear that the quotient isn't so small after all.  It's another thing to experiment with - I hoped that since that paper is now fairly old (In computer hardware terms) that these kinds of tricks would no longer be required but we'll see!

NAThompson commented:
> I was getting screwy/slow results from whatever run went first, so it was just a safeguard against deluding myself ;)

I've seen that too; the CPU heats up and then downclocks in the middle of the run. At least on Linux you can stop this with

```
$ sudo cpupower frequency-set --governor performance
```

jzmaddock commented:
I think this is about ready now: fixed a few bugs, added much stricter assertions and tests, and implemented the double-digit version of the algorithm.

Performance relative to GMP first:



GCC-10: |   |  
-- | -- | --
  |   |  
Bits | Old | New
512 | 3.00 | 1.72
1024 | 4.06 | 1.71
2048 | 5.39 | 1.71
4096 | 7.84 | 1.69
8192 | 11.51 | 1.65
16384 | 16.00 | 1.64
32768 | 22.30 | 1.73


MSVC-14.2: |   |  
-- | -- | --
  |   |  
Bits | Old | New
512 | 3.05 | 1.34
1024 | 3.87 | 1.40
2048 | 5.07 | 1.60
4096 | 6.95 | 1.43
8192 | 8.64 | 1.46
16384 | 10.49 | 1.36
32768 | 11.85 | 1.33

Raw performance data:

```
GCC-10:
Run on (8 X 1992 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 262K (x4)
  L3 Unified 8388K (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_gcd_current<cpp_int>/512         5.19 ms         5.19 ms         1318
BM_gcd_current<cpp_int>/1024        11.1 ms         11.1 ms          631
BM_gcd_current<cpp_int>/2048        24.3 ms         24.3 ms          289
BM_gcd_current<cpp_int>/4096        53.9 ms         54.0 ms          128
BM_gcd_current<cpp_int>/8192         127 ms          127 ms           55
BM_gcd_current<cpp_int>/16384        339 ms          339 ms           21
BM_gcd_current<cpp_int>/32768       1012 ms         1011 ms            7
BM_gcd_current<cpp_int>_BigO        0.97 N^2        0.96 N^2
BM_gcd_current<cpp_int>_RMS           19 %            19 %
BM_gcd_old<cpp_int>/512             9.63 ms         9.62 ms          723
BM_gcd_old<cpp_int>/1024            26.3 ms         26.4 ms          268
BM_gcd_old<cpp_int>/2048            76.6 ms         76.6 ms           91
BM_gcd_old<cpp_int>/4096             250 ms          250 ms           28
BM_gcd_old<cpp_int>/8192             886 ms          885 ms            8
BM_gcd_old<cpp_int>/16384           3317 ms         3312 ms            2
BM_gcd_old<cpp_int>/32768          12990 ms        13000 ms            1
BM_gcd_old<cpp_int>_BigO           12.12 N^2       12.13 N^2
BM_gcd_old<cpp_int>_RMS                2 %             2 %
BM_gcd_current<mpz_int>/512         3.02 ms         3.02 ms         2333
BM_gcd_current<mpz_int>/1024        6.47 ms         6.48 ms         1093
BM_gcd_current<mpz_int>/2048        14.2 ms         14.2 ms          498
BM_gcd_current<mpz_int>/4096        31.9 ms         31.9 ms          220
BM_gcd_current<mpz_int>/8192        76.9 ms         76.9 ms           91
BM_gcd_current<mpz_int>/16384        207 ms          207 ms           35
BM_gcd_current<mpz_int>/32768        583 ms          583 ms           12
BM_gcd_current<mpz_int>_BigO        0.56 N^2        0.56 N^2
BM_gcd_current<mpz_int>_RMS           22 %            22 %
BM_gcd_current<mpz_int>/512         3.01 ms         3.01 ms         2321

MSVC-14.2:
Run on (8 X 1992 MHz CPU s)
CPU Caches:
  L1 Data 32K (x4)
  L1 Instruction 32K (x4)
  L2 Unified 262K (x4)
  L3 Unified 8388K (x1)
------------------------------------------------------------------------
Benchmark                              Time             CPU   Iterations
------------------------------------------------------------------------
BM_gcd_old<cpp_int>/512             14.0 ms         14.0 ms          503
BM_gcd_old<cpp_int>/1024            38.5 ms         38.5 ms          181
BM_gcd_old<cpp_int>/2048             115 ms          115 ms           61
BM_gcd_old<cpp_int>/4096             375 ms          376 ms           19
BM_gcd_old<cpp_int>/8192            1314 ms         1316 ms            5
BM_gcd_old<cpp_int>/16384           4915 ms         4914 ms            2
BM_gcd_old<cpp_int>/32768          18787 ms        18781 ms            1
BM_gcd_old<cpp_int>_BigO           17.55 N^2       17.55 N^2
BM_gcd_old<cpp_int>_RMS                3 %             3 %
BM_gcd_current<cpp_int>/512         6.11 ms         6.11 ms         1120
BM_gcd_current<cpp_int>/1024        13.9 ms         13.9 ms          503
BM_gcd_current<cpp_int>/2048        36.3 ms         36.3 ms          211
BM_gcd_current<cpp_int>/4096        79.5 ms         79.4 ms           76
BM_gcd_current<cpp_int>/8192         224 ms          223 ms           33
BM_gcd_current<cpp_int>/16384        635 ms          633 ms           10
BM_gcd_current<cpp_int>/32768       2114 ms         2109 ms            3
BM_gcd_current<cpp_int>_BigO        2.00 N^2        1.99 N^2
BM_gcd_current<cpp_int>_RMS           13 %            13 %
BM_gcd_old<cpp_int>/512             13.9 ms         13.9 ms          503
BM_gcd_old<cpp_int>/1024            38.5 ms         38.5 ms          182
BM_gcd_old<cpp_int>/2048             116 ms          115 ms           61
BM_gcd_old<cpp_int>/4096             387 ms          387 ms           19
BM_gcd_old<cpp_int>/8192            1320 ms         1322 ms            5
BM_gcd_old<cpp_int>/16384           4902 ms         4898 ms            2
BM_gcd_old<cpp_int>/32768          18798 ms        18797 ms            1
BM_gcd_old<cpp_int>_BigO           17.56 N^2       17.56 N^2
BM_gcd_old<cpp_int>_RMS                3 %             3 %
BM_gcd_current<mpz_int>/512         4.55 ms         4.55 ms         1508
BM_gcd_current<mpz_int>/1024        9.95 ms         9.96 ms          700
BM_gcd_current<mpz_int>/2048        22.7 ms         22.7 ms          309
BM_gcd_current<mpz_int>/4096        55.7 ms         55.7 ms          124
BM_gcd_current<mpz_int>/8192         152 ms          153 ms           46
BM_gcd_current<mpz_int>/16384        467 ms          467 ms           15
BM_gcd_current<mpz_int>/32768       1587 ms         1586 ms            4
BM_gcd_current<mpz_int>_BigO        1.50 N^2        1.50 N^2
BM_gcd_current<mpz_int>_RMS           11 %            11 %
```

NAThompson commented on include/boost/multiprecision/cpp_int/misc.hpp:
```diff
@@ -495,14 +495,491 @@ eval_gcd(
+// __int128.  For 64 bit limbs it reduces the size of U by about 62 bits per call.
+//
+// The complexity of the algorithm given by Sorenson is roughly O(ln^2(N)) for
+// two N bit numbers.
```
Wait, isn't the benchmark showing O(N^2) complexity for two N-bit numbers?

jzmaddock replied:
It alternates from run to run between O(n^2) and O(NlnN), this is true of GMP also.  I don't think that either complexity is actually correct.

NAThompson commented on performance/gcd_bench.cpp:
```diff
@@ -0,0 +1,215 @@
+BENCHMARK_TEMPLATE(BM_gcd_current, cpp_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_old, cpp_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_current, mpz_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
+BENCHMARK_TEMPLATE(BM_gcd_current, mpz_int)->RangeMultiplier(2)->Range(lower_range, upper_range)->Unit(benchmark::kMillisecond)->Complexity();
```
I think now that this is finished, we can remove the duplicated performance tests since you'd demonstrated an order of magnitude speedup. It's a pretty unambiguous win now!

jzmaddock replied:
Nod.

NAThompson commented on performance/gcd_bench.cpp:
```diff
@@ -0,0 +1,215 @@
+
```
Do we need the copyright notice here?

jzmaddock replied:
Yes, just seen that in the failed inspection report, I'll wait for the other CI tests to cycle before making a few cosmetic changes like this.

NAThompson commented on performance/gcd_bench.cpp:
```diff
@@ -0,0 +1,215 @@
+   for (auto _ : state)
+   {
+      for (unsigned i = 0; i < a.size(); ++i)
+         eval_gcd(c[i].backend(), a[i].backend(), b[i].backend());
```
Wait, why are you doing an eval_gcd on the entire vector? Why not just do it on two single numbers?

jzmaddock replied:
GCD performance is highly dependent on the inputs, you want to try and get a decent average.  To pick a trivial example, if one of the numbers has a lot of consecutive zero bits, then binary GCD would look way better than it actually is.

NAThompson commented:
@jzmaddock : I wasn't able to give a super meaningful review, but this is awesome. I also ran it under `-fsanitize=address -fsanitize=undefined` and found nothing.
